staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go

```diff
@@ -0,0 +1,109 @@
+package storage
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/component-base/version"
+)
+
+type ResourceSpec struct {
+	// Version indicates the Kubernetes version from which this GVR is decodable
+	K8sVersion semver.Version
+
+	// Version indicates the resource version for which support was added
+	Version string
+}
+
+type ResourceHistory = map[schema.GroupResource][]ResourceSpec
+
+type CompatibilityResourceEncodingConfig struct {
+	binaryVersion        semver.Version
+	compatibilityVersion semver.Version
+	versionSkew          uint8
+	resourceHistory      ResourceHistory
+}
+
+var _ ResourceEncodingConfig = &CompatibilityResourceEncodingConfig{}
+
+var v1_26 = semver.MustParse("v1.26.0")
+var v1_27 = semver.MustParse("v1.27.0")
+
+func NewCompatibilityResourceConfig() *CompatibilityResourceEncodingConfig {
+	bv := getBinaryVersion()
+	return &CompatibilityResourceEncodingConfig{
+		binaryVersion:        bv,
+		compatibilityVersion: bv,
+		resourceHistory:      ResourceHistory{
+			//!TODO: populate with history
+			// schema.GroupResource{Group: "authorization.k8s.io", Resource: "localsubjectaccessreviews"}: {
```

Comment:
The resource history is intended to inform which k8s version a given GVR became supported by k8s. Right now I have it empty. Curious for comments about:

1. The ResourceHistory data structure for tracking when GVRs were supported
2. Reliable ways to keep it populated without manual labor

Reply:
test/integration/etcd/data.go and staging/src/k8s.io/api/testdata are two places where every group/version/kind we know about in kube-apiserver has an associated entry

We make use of the skew fixture data in staging/src/k8s.io/api/testdata in the etcd test to make sure we don't switch the storage version to a version the previous server did not understand if the previous server understood any non-alpha versions:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/etcd/etcd_storage_path_test.go#L166-L207

I don't know that I'd copy those approaches exactly, but something along those lines could be an idea